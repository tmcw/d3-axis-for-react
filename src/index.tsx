import React from "react";
import type { AxisScale, AxisDomain } from "d3-axis";

function identity(x: any) {
  return x;
}

/**
 * Instead of a component for each orientation (like AxisLeft, AxisRight),
 * we provide a value from this Orient object. Provide a value, like
 * Orient.left, to the `orient` prop of the Axis component
 * to place the axis on the left.
 */
export enum Orient {
  top = 1,
  right = 2,
  bottom = 3,
  left = 4,
}

function translateX(x: number) {
  return "translate(" + x + ",0)";
}

function translateY(y: number) {
  return "translate(0," + y + ")";
}

/**
 * The axis component. This renders an axis, within a
 * `g` element, for use in a chart.
 */
export const Axis = <Domain extends AxisDomain>({
  scale,
  ticks,
  tickArguments = [],
  tickValues = null,
  tickFormat = null,
  tickSize,
  tickSizeInner = 6,
  tickSizeOuter = 6,
  tickPadding = 3,
  tickTextProps = {},
  tickLineProps = {},
  domainPathProps = {},
  orient = Orient.bottom,
  offset = typeof window !== "undefined" && window.devicePixelRatio > 1
    ? 0
    : 0.5,
}: {
  /** An initialized d3 scale object, like a d3.linearScale */
  scale: AxisScale<Domain>;
  ticks?: any[];
  tickArguments?: any[];
  tickValues?: any[] | null;
  tickFormat?: any;
  tickSize?: number;
  tickSizeInner?: number;
  tickSizeOuter?: number;
  tickPadding?: number;
  /** Additional attributes to add to tick text elements, or null to omit */
  tickTextProps?: React.SVGProps<SVGTextElement> | null;
  /** Additional attributes to add to tick line elements, or null to omit */
  tickLineProps?: React.SVGProps<SVGLineElement> | null;
  /** Additional attributes to the domain path, or null to omit */
  domainPathProps?: React.SVGProps<SVGPathElement> | null;
  offset?: number;
  orient?: Orient;
}) => {
  if (tickSize) {
    tickSizeInner = tickSize;
    tickSizeOuter = tickSize;
  }

  if (ticks) {
    tickArguments = ticks;
  }

  function number(scale: AxisScale<Domain>) {
    return (d: any) => {
      const value = scale(d);
      return value === undefined ? 0 : +value;
    };
  }

  function center(scale: AxisScale<Domain>, offset: number) {
    if (scale.bandwidth) {
      offset = Math.max(0, scale.bandwidth() - offset * 2) / 2;
    }
    if ((scale as any).round()) offset = Math.round(offset);
    return (d: Domain) => {
      const value = scale(d);
      return value === undefined ? 0 : value + offset;
    };
  }

  const k = orient === Orient.top || orient === Orient.left ? -1 : 1,
    x = orient === Orient.left || orient === Orient.right ? "x" : "y",
    transform =
      orient === Orient.top || orient === Orient.bottom
        ? translateX
        : translateY;

  // Rendering
  const values =
      tickValues == null
        ? (scale as any).ticks
          ? (scale as any).ticks.apply(scale, tickArguments)
          : scale.domain()
        : tickValues,
    format =
      tickFormat == null
        ? "tickFormat" in scale
          ? (scale as any).tickFormat.apply(scale, tickArguments)
          : identity
        : tickFormat,
    spacing = Math.max(tickSizeInner, 0) + tickPadding,
    range = scale.range(),
    range0 = +range[0] + offset,
    range1 = +range[range.length - 1] + offset,
    position = (scale.bandwidth ? center : number)(scale.copy(), offset);

  const domainPath =
    orient === Orient.left || orient === Orient.right
      ? tickSizeOuter
        ? "M" +
          k * tickSizeOuter +
          "," +
          range0 +
          "H" +
          offset +
          "V" +
          range1 +
          "H" +
          k * tickSizeOuter
        : "M" + offset + "," + range0 + "V" + range1
      : tickSizeOuter
      ? "M" +
        range0 +
        "," +
        k * tickSizeOuter +
        "V" +
        offset +
        "H" +
        range1 +
        "V" +
        k * tickSizeOuter
      : "M" + range0 + "," + offset + "H" + range1;

  const lineProps = {
    [x + "2"]: k * tickSizeInner,
  };

  const textProps = {
    [x]: k * spacing,
  };

  return (
    <g>
      {values.map((tick: any, i: number) => (
        <g
          className="tick"
          key={i}
          transform={transform(position(tick) + offset)}
        >
          {tickLineProps && (
            <line stroke="currentColor" {...lineProps} {...tickLineProps} />
          )}
          {tickTextProps && (
            <text
              fill="currentColor"
              dy={
                orient === Orient.top
                  ? "0em"
                  : orient === Orient.bottom
                  ? "0.71em"
                  : "0.32em"
              }
              fontSize="10"
              fontFamily="sans-serif"
              textAnchor={
                orient === Orient.right
                  ? "start"
                  : orient === Orient.left
                  ? "end"
                  : "middle"
              }
              {...textProps}
              {...tickTextProps}
            >
              {format(tick)}
            </text>
          )}
        </g>
      ))}
      {domainPathProps && (
        <path
          className="domain"
          stroke="currentColor"
          fill="transparent"
          d={domainPath}
          {...domainPathProps}
        />
      )}
    </g>
  );
};
