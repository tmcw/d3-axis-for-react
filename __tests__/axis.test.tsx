import React from "react";
import renderer from "react-test-renderer";
import { Axis, Orient } from "../src/index";
import { scaleLinear } from "d3";
import { test, expect } from "vitest";

test("axis with default options", async () => {
  const scale = scaleLinear().domain([0, 1]).range([0, 1]);
  const component = renderer.create(<Axis<number> scale={scale} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("axis with top orientation", async () => {
  const scale = scaleLinear().domain([0, 1]).range([0, 1]);
  const component = renderer.create(
    <Axis<number> orient={Orient.top} scale={scale} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("axis with domain and path omitted", async () => {
  const scale = scaleLinear().domain([0, 1]).range([0, 1]);
  const component = renderer.create(
    <Axis<number>
      orient={Orient.top}
      scale={scale}
      tickLineProps={null}
      tickTextProps={null}
      domainPathProps={null}
    />
  );
  let tree = component.toJSON();
  expect(JSON.stringify(tree)).not.toMatch(/text/g);
  expect(tree).toMatchSnapshot();
});

test("axis with custom text props", async () => {
  const scale = scaleLinear().domain([0, 1]).range([0, 1]);
  const component = renderer.create(
    <Axis<number>
      orient={Orient.top}
      scale={scale}
      tickTextProps={{
        fontSize: "20",
      }}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
