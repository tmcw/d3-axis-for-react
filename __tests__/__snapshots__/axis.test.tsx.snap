// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`axis with custom text props 1`] = `
<g>
  <g
    className="tick"
    transform="translate(0.5,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.0
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.6,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.1
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.7,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.2
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.8,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.3
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.9,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.4
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.5
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.1,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.6
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.2,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.7
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.3,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.8
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.4,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      0.9
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.5,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="20"
      textAnchor="middle"
      y={-9}
    >
      1.0
    </text>
  </g>
  <path
    className="domain"
    d="M0.5,-6V0.5H1.5V-6"
    fill="transparent"
    stroke="currentColor"
  />
</g>
`;

exports[`axis with default options 1`] = `
<g>
  <g
    className="tick"
    transform="translate(0.5,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.0
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.6,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.1
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.7,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.2
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.8,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.3
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.9,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.4
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.5
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.1,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.6
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.2,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.7
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.3,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.8
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.4,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.9
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.5,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      1.0
    </text>
  </g>
  <path
    className="domain"
    d="M0.5,6V0.5H1.5V6"
    fill="transparent"
    stroke="currentColor"
  />
</g>
`;

exports[`axis with domain and path omitted 1`] = `
<g>
  <g
    className="tick"
    transform="translate(0.5,0)"
  />
  <g
    className="tick"
    transform="translate(0.6,0)"
  />
  <g
    className="tick"
    transform="translate(0.7,0)"
  />
  <g
    className="tick"
    transform="translate(0.8,0)"
  />
  <g
    className="tick"
    transform="translate(0.9,0)"
  />
  <g
    className="tick"
    transform="translate(1,0)"
  />
  <g
    className="tick"
    transform="translate(1.1,0)"
  />
  <g
    className="tick"
    transform="translate(1.2,0)"
  />
  <g
    className="tick"
    transform="translate(1.3,0)"
  />
  <g
    className="tick"
    transform="translate(1.4,0)"
  />
  <g
    className="tick"
    transform="translate(1.5,0)"
  />
</g>
`;

exports[`axis with top orientation 1`] = `
<g>
  <g
    className="tick"
    transform="translate(0.5,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.0
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.6,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.1
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.7,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.2
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.8,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.3
    </text>
  </g>
  <g
    className="tick"
    transform="translate(0.9,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.4
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.5
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.1,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.6
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.2,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.7
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.3,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.8
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.4,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.9
    </text>
  </g>
  <g
    className="tick"
    transform="translate(1.5,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      1.0
    </text>
  </g>
  <path
    className="domain"
    d="M0.5,-6V0.5H1.5V-6"
    fill="transparent"
    stroke="currentColor"
  />
</g>
`;
